### ISSUE TITLE
Weird behavior in remix-tests when the bytecode is of a certain size

### ISSUE THREAD

**Issue by marco2357** [ethereum/solidity#11378]
## Description

I just updated my project to Solidity 0.8.4. I changed this debugging/helper function:

    function ok() public view returns (string memory) {
        return string(abi.encodePacked(
            foo1 ? "XXXXXXXXXXXX" : "",
            foo2 ? "XXXXX" : "XXXX",
            "XXXXXXXX",
            foo3 ? "" : "XXXXXXXXXXXXXXX",
            "XXX",
            foo4,
            "X"
         ));
    }

to:

    function bad() public view returns (string memory) {
        return string(bytes.concat(
            bytes(foo1 ? "XXXXXXXXXXXX" : ""),
            bytes(foo2 ? "XXXXX" : "XXXX"),
            "XXXXXXXX",
            bytes(foo3 ? "" : "XXXXXXXXXXXXXXX"),
            "XXX",
            bytes(foo4),
            "X"
        ));
    }

When executing my contract now, **without** calling the ok/bad function at all, it fails reproducible at a different/independent location. Adding a pointless `Assert.equal(var1, var2)` at that location makes the error "go away". It looks to me like the byte code is somehow corrupted. I went back to abi.encodePacked() since this seems to work fine and I don't need the additional bytes() casts.

## Environment

- Compiler version: 0.8.4
- Target EVM version (as per compiler settings): remix-tests default
- Framework/IDE (e.g. Truffle or Remix): remix-tests (local execution)
- EVM execution environment / backend / blockchain client: remix-tests default (JavaScript?)
- Operating system: Linux

## Steps to Reproduce

The code is ~1k locs. Minimizing it while keeping the bug reproducible would take me quite a while. Removing one X from one of the strings already makes the error "go away". So I hope I can avoid the minimizing work by first checking a couple of other points:

- Is there an obvious problem with/an obvious error in the bad() code? Maybe I wrote invalid code that compiles to invalid byte code?
- Maybe someone knowing the compiler well directly has an intuition what might go wrong?
- Unfortunately my only idea left is then minimizing the code. Since I'm short on time and `abi.encodePacked()` works, I'm somewhat reluctant.

**Comment by axic** [comment]
Sorry for the delay, but any chance you can share the source code? Obviously a smaller reproducible snippet is better, but it is really hard to do anything without seeing it happen in person.

If you could share two versions, one which produces the issue and which doesn't that would be the best.

**Comment by chriseth** [comment]
To me it's not clear what the actual error is you are describing. Your conclusion is "byte code corruption", but the error is " it fails reproducible at a different/independent location". What do you mean by "reproducible"?

Note that `bytes.concat` does not generate exactly the same bytecode as `abi.encodePacked`.

**Comment by marco2357** [comment]
I attached a "minimal" example: [problem.tar.gz](https://github.com/ethereum/solidity/files/6566358/problem.tar.gz)

`tar xfz problem.tar.gz`
`cd problem`

`remix-tests --optimize true --runs 200 test/SellAtMarket_test.sol`
Error: The contract code couldn't be stored, please check your gas limit.

Increase the gas limit the ugly way (I didn't find a better way):
`vi node-v14.16.0-linux-x64/lib/node_modules/@remix-project/remix-tests/node_modules/ethereumjs-vm/dist/evm/message.js`
Add `opts.gasLimit.words[0] *= 20;` to the top of `function Message(opts) {`.

`remix-tests --optimize true --runs 200 test/SellAtMarket_test.sol`
2 passing (10.86s)

`remix-tests --optimize true --runs 100 test/SellAtMarket_test.sol`
error: Transaction has been reverted by the EVM

I couldn't minimize the code in problems.tar.gz any more. Any lines I removed made the error disappear. Also, changing gasLimit factor to e.g. 10 makes the error disappear.
So the problem might not be caused by bytes.concat(). My change to remix-tests might be the cause or maybe there is a problem with the optimization.

Hope this helps. Feel free to close the issue if you think it's caused by my remix-tests change.

**Comment by cameel** [comment]
> `remix-tests --optimize true --runs 200 test/SellAtMarket_test.sol`
> 2 passing (10.86s)
>
> `remix-tests --optimize true --runs 100 test/SellAtMarket_test.sol`
> error: Transaction has been reverted by the EVM

The fact that just changing the value of `--runs` makes the test pass/fail is very weird. I managed to reproduce it and spent quite a bit of time trying to figure out what's going on and looks like it's not related to `bytes.concat()`. It fails only when the bytecode has a very specific length. If you replace the `bytecode.concat()` call in your `OpenEndKnockOutWarrant.description()` function with an instruction returning a string of exactly 501 characters (i.e. exactly as much bytecode as your `bytes.concat()` call takes), it behaves the same way. Adding/removing even a single character from the string makes the failure go away.

I'm not sure if it's a problem in the compiler or remix-tests. It's technically possible that something wrong happens in `solc` when the code size is exactly at the size limit (e.g. it gets truncated or something) but it's also very possible that there's something wrong with the transaction remix-tests uses to deploy it. There might also be some off-by-one error in some check along the way.

### Repro with remix-tests
This how I reproduced it locally:
``` bash
npm install @remix-project/remix-tests
npm install tslib
sed -i 's/\(function Message(opts) {\)/\1opts.gasLimit.words[0] *= 20;/' node_modules/ethereumjs-vm/dist/evm/message.js

tar xfv problem.tar.gz
cd problem/
npx remix-tests --compiler 0.8.4 --optimize true --runs 200 test/SellAtMarket_test.sol
npx remix-tests --compiler 0.8.4 --optimize true --runs 100 test/SellAtMarket_test.sol
```

Also, here's the 501-character-long string I used:

``` solidity
return
    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"
    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"
    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"
    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"
    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"

    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"
    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"
    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"
    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"
    "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX" "XXXXXXXXXX"

    "X";
```

#### Test output
This is with `bytes.concat()` already replaced with the string I gave above.
##### `--runs 200`
```
[19:59:14] info: Compiler version set to 0.8.4. Latest version is 0.8.4
[19:59:14] info: Optimization is enabled
[19:59:14] info: Runs set to 200
Loading remote solc version v0.8.4+commit.c7e474f2 ...
'creation of library remix_tests.sol:Assert pending...'

        ◼  TestSellAtMarket
        ✓  Test regular
        ✓  Test small


2 passing (6.313000000000001s)
```

##### `--runs 100`
```
[19:57:19] info: Compiler version set to 0.8.4. Latest version is 0.8.4
[19:57:19] info: Optimization is enabled
[19:57:19] info: Runs set to 100
Loading remote solc version v0.8.4+commit.c7e474f2 ...
'creation of library remix_tests.sol:Assert pending...'

        ◼  TestSellAtMarket
        ✓  Test regular
        ✘  Test small


1 passing (4.781000000000001s)
1 failing

  1) TestSellAtMarket: Test small

         error: Transaction has been reverted by the EVM:
{
  "transactionHash": "0xf1a1d6b6841ccc8a5455829ea453e99e65340a0b0c5f13e2f7cf45f90c3a1015",
  "transactionIndex": 0,
  "blockHash": "0x69f7b7a30c97eb7f1bbea2c8e72b43677e395381e30f78d2edcc8411852b36b9",
  "blockNumber": 4,
  "gasUsed": 5000000,
  "cumulativeGasUsed": 5000000,
  "status": false,
  "to": "0xf8e81D47203A594245E36C48e151709F0C19fBe8",
  "events": {}
}
         expected value to be undefined to: undefined
         returned: undefined
```

### Repro with `solc`
To reproduce with solc, something like this should work (executed inside `problem/`):
``` bash
curl -L https://raw.githubusercontent.com/ethereum/remix-project/master/libs/remix-tests/sol/tests.sol.ts --output remix_tests.sol
sed -i 's/module.exports = `//' remix_tests.sol
sed -i 's/`//' remix_tests.sol
solc test/SellAtMarket_test.sol --allow-paths . --optimize --optimize-runs 100 --bin --libraries remix_tests.sol:Assert=0x1234567890123456789012345678901234567890
```

The compiler does not seem to have any problems generating code from the provided source though. Both at its current length and when I add/remove parts. I do not see anything interesting in the `--bin` output and the differences in `--asm` seem to be just offsets and data hashes. The output between `--runs 100` and `--runs 200` does differ quite a bit but it's hard to investigate since it's just all marked as different after a certain point.

I'm not sure I'm getting exactly the same content as remix though. `remix_tests.sol` is generated by remix-tests on the fly so it might not be the same exact version and the tool might be doing some extra modifications along the way. Also, my address for the `Assert` library is just a random value, not the same thing remix-test passes to the compiler.

**Comment by cameel** [comment]
Issue reported in Remix: https://github.com/ethereum/remix-project/issues/1246. Hopefully they can help debug this.

**Comment by cameel** [comment]
Looks the issue in remix-tests has been fixed (https://github.com/ethereum/remix-project/pull/1249/files).

@marco2357 Could you check if that resolves your original problem?

**Comment by marco2357** [comment]
Yes, I can confirm that this problem is solved in remix-tests 0.2.0 (remix 0.12.0). I'm closing the issue.
Thank you Kamil!
